codeworld-error-sanitizer/src/RegexShim.hs

```diff
@@ -38,23 +39,45 @@ replace regex replacement str = textFromJSString $
 
 #else
 
-import Data.Array (elems)
+import Data.Array (bounds, elems, (!))
+import Data.Char (isDigit)
 import Data.List
 import Data.Monoid
 import qualified Data.Text as T
+import Data.Text.Read (decimal)
 import Text.Regex.Base
 import Text.Regex.TDFA
 import Text.Regex.TDFA.Text
 
+getInserts :: Text -> [(Int, (Int, Int))]
+getInserts replacement = case replacement =~ ("[\\][0-9]" :: Text) :: [MatchArray] of
+    [] -> []
+    groups -> fmap processInsertion groups
+    where processInsertion group =
+              case decimal $ T.take len $ T.drop (start + 1) replacement of
+                  Right (index, _) -> (index, (start, len))
+                  Left _ -> error $ T.unpack $ "Can not parse replacement groups in "
+                                  <> replacement -- should never happen beause handled
+                                                 -- by regex matching above
+              where (start, len) = group ! 0
+
+applyInserts :: Text -> Text -> MatchArray -> [(Int, (Int, Int))] -> Text
+applyInserts str replacement matches ((index, (rstart, rlen)):inserts) =
+    let pre = T.take rstart replacement
+        rest = T.drop (rstart + rlen) replacement
+        (sstart, slen) = matches ! index
+        source = T.take slen $ T.drop sstart str
+    in pre <> source <> applyInserts str rest matches inserts
+applyInserts _ replacement _ [] = replacement
+
 replace :: Text -> Text -> Text -> Text
-replace regex replacement str =
-    let parts = concatMap elems $ (str =~ regex :: [MatchArray])
-    in foldl replaceOne str (reverse parts)
-  where
-    replaceOne :: Text -> (Int, Int) -> Text
-    replaceOne str (start, len) = pre <> replacement <> post
-      where
-        pre = T.take start str
-        post = T.drop (start + len) str
+replace regex replacement str = case str =~ regex of
```

Comment:
This only matches one occurrence of the target string.  That's different from the current behavior, which matches all occurrences.  I think you still need the fold to handle that.  (This might need more fiddling with indices, though.  Haven't thought it through.)

Reply:
Thanks, fixed.